File: stvs/src/main/java/edu/kit/iti/formal/stvs/logic/io/xml/XmlExporter.java
package edu.kit.iti.formal.stvs.logic.io.xml;

import edu.kit.iti.formal.stvs.io._1.ObjectFactory;
import edu.kit.iti.formal.stvs.logic.io.ExportException;
import edu.kit.iti.formal.stvs.logic.io.Exporter;
import org.w3c.dom.Document;
import org.w3c.dom.Node;

import javax.xml.bind.JAXBContext;
import javax.xml.bind.JAXBElement;
import javax.xml.bind.JAXBException;
import javax.xml.bind.Marshaller;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.transform.OutputKeys;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerException;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.StringWriter;

/**
 * Common superclass for all Exporters that export to xml.
 *
 * @author Benjamin Alt
 */
public abstract class XmlExporter<F> implements Exporter<F> {

    public static final Class<ObjectFactory> OF_STVS = ObjectFactory.class;
    public static final Class<edu.kit.iti.formal.exteta_1.ObjectFactory> OF_EXTETA = edu.kit.iti.formal.exteta_1.ObjectFactory.class;
    public static final Class<edu.kit.iti.formal.exteta_1_0.report.ObjectFactory> OF_EXTETA_REPORT = edu.kit.iti.formal.exteta_1_0.report.ObjectFactory.class;

    //public static final String NAMESPACE = "http://formal.iti.kit.edu/stvs/io/1.0/";

    /**
     * Marshals a {@link JAXBElement} to a xml node.
     *
     * @param element The element that should be marshaled
     * @param clazz class of the ObjectFactory
     * @return Node representing the input {@code element}
     * @throws ExportException Exception while exporting
     */
    protected static Node marshalToNode(JAXBElement element, Class clazz)
            throws ExportException {
        try {
            DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
            DocumentBuilder db = dbf.newDocumentBuilder();
            Document document = db.newDocument();
            JAXBContext context = JAXBContext.newInstance(clazz);
            Marshaller marshaller = context.createMarshaller();
            marshaller.setProperty("jaxb.formatted.output", Boolean.TRUE);
            marshaller.marshal(element, document);
            return document.getDocumentElement();
        } catch (ParserConfigurationException | JAXBException e) {
            throw new ExportException(e);
        }
    }

    /**
     * Exports an Object as xml.
     *
     * @param source Object to export
     * @return The output xml is written to this stream
     * @throws ExportException Exception while exporting
     */
    public ByteArrayOutputStream export(F source) throws ExportException {
        Node xmlNode = exportToXmlNode(source);
        StringWriter writer = new StringWriter();
        try {
            Transformer transformer = TransformerFactory.newInstance().newTransformer();
            transformer.setOutputProperty(OutputKeys.INDENT, "yes");
            transformer.transform(new DOMSource(xmlNode), new StreamResult(writer));
            String xmlString = writer.toString();
            ByteArrayOutputStream stream = new ByteArrayOutputStream();
            stream.write(xmlString.getBytes("utf-8"));
            return stream;
        } catch (TransformerException | IOException e) {
            throw new ExportException(e);
        }
    }

    /**
     * Must be implemented by subclasses. This method must provide the logic to convert the given
     * {@code source} object into a xml {@link Node}
     *
     * @param source Element that should be converted
     * @return Xml node
     * @throws ExportException Exception while exporting
     */
    public abstract Node exportToXmlNode(F source) throws ExportException;
}

File: stvs/src/main/kotlin/edu/kit/iti/formal/stvs/logic/io/xml/XmlExporter.kt
/* *****************************************************************
 * This file belongs to verifaps-lib (https://verifaps.github.io).
 * SPDX-License-Header: GPL-3.0-or-later
 * 
 * This program isType free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program isType distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a clone of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * *****************************************************************/
package edu.kit.iti.formal.stvs.logic.io.xml

import edu.kit.iti.formal.stvs.logic.io.ExportException
import edu.kit.iti.formal.stvs.logic.io.Exporter
import org.jdom2.Document
import org.jdom2.Element
import org.jdom2.Namespace
import org.jdom2.output.Format
import org.jdom2.output.XMLOutputter
import org.w3c.dom.Node
import java.io.OutputStream
import javax.xml.transform.TransformerException

/**
 * Common superclass for all Exporters that export to xml.
 *
 * @author Benjamin Alt
 */
abstract class XmlExporter<F> : Exporter<F> {
    /**
     * Exports an Object as xml.
     *
     * @param source Object to export
     * @return The output xml is written to this stream
     * @throws ExportException Exception while exporting
     */
    @Throws(ExportException::class)
    override fun export(source: F, target: OutputStream) {
        val xmlNode = exportToXmlNode(source)
        try {
            val xmlOutputter = XMLOutputter(
                Format.getPrettyFormat()
                    .setEncoding("UTF-8")
                    .setOmitEncoding(false)
                    .setOmitDeclaration(false),
            )
            xmlOutputter.output(Document(xmlNode), target)
        } catch (e: TransformerException) {
            throw ExportException(e)
        }
    }

    /**
     * Must be implemented by subclasses. This method must provide the logic to convert the given
     * `source` object into a xml [Node]
     *
     * @param source Element that should be converted
     * @return Xml node
     * @throws ExportException Exception while exporting
     */
    @Throws(ExportException::class)
    abstract fun exportToXmlNode(source: F): Element
}

const val NAMESPACE_URI = "http://formal.iti.kit.edu/stvs/io/1.0/"
val NAMESPACE = Namespace.getNamespace(NAMESPACE_URI)